From 9496a622c0574470069f0e8fc5f5f550d6cfe9d0 Mon Sep 17 00:00:00 2001
From: Liu Ying <victor.liu@nxp.com>
Date: Fri, 31 Aug 2018 10:42:19 +0800
Subject: [PATCH 4501/5242] MLK-19411 gpu: imx: dpu: units: Fix bailout path
 of dpu_*_get()

commit  a563842162ffc809882dc7c07683ac5a845b13df from
https://source.codeaurora.org/external/imx/linux-imx.git

If a DPU unit is already in use, the bailout path of dpu_{unit}_get()
would wrongly dereference the pointer of ERR_PTR(-EBUSY).  This patch
fixes this issue.

Signed-off-by: Liu Ying <victor.liu@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/imx/dpu/dpu-constframe.c  |    8 ++++----
 drivers/gpu/imx/dpu/dpu-disengcfg.c   |    8 ++++----
 drivers/gpu/imx/dpu/dpu-extdst.c      |    8 ++++----
 drivers/gpu/imx/dpu/dpu-fetchdecode.c |    6 +++---
 drivers/gpu/imx/dpu/dpu-fetcheco.c    |    6 +++---
 drivers/gpu/imx/dpu/dpu-fetchlayer.c  |    6 +++---
 drivers/gpu/imx/dpu/dpu-fetchwarp.c   |    6 +++---
 drivers/gpu/imx/dpu/dpu-framegen.c    |    6 +++---
 drivers/gpu/imx/dpu/dpu-hscaler.c     |    8 ++++----
 drivers/gpu/imx/dpu/dpu-layerblend.c  |    6 +++---
 drivers/gpu/imx/dpu/dpu-tcon.c        |    8 ++++----
 drivers/gpu/imx/dpu/dpu-vscaler.c     |    6 +++---
 12 files changed, 41 insertions(+), 41 deletions(-)

diff --git a/drivers/gpu/imx/dpu/dpu-constframe.c b/drivers/gpu/imx/dpu/dpu-constframe.c
index d014830..57206bb 100644
--- a/drivers/gpu/imx/dpu/dpu-constframe.c
+++ b/drivers/gpu/imx/dpu/dpu-constframe.c
@@ -1,6 +1,6 @@
 /*
  * Copyright (C) 2016 Freescale Semiconductor, Inc.
- * Copyright 2017 NXP
+ * Copyright 2017-2018 NXP
  *
  * This program is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License as published by the
@@ -149,12 +149,12 @@ struct dpu_constframe *dpu_cf_get(struct dpu_soc *dpu, int id)
 	mutex_lock(&cf->mutex);
 
 	if (cf->inuse) {
-		cf = ERR_PTR(-EBUSY);
-		goto out;
+		mutex_unlock(&cf->mutex);
+		return ERR_PTR(-EBUSY);
 	}
 
 	cf->inuse = true;
-out:
+
 	mutex_unlock(&cf->mutex);
 
 	return cf;
diff --git a/drivers/gpu/imx/dpu/dpu-disengcfg.c b/drivers/gpu/imx/dpu/dpu-disengcfg.c
index 007da25..76e2188 100644
--- a/drivers/gpu/imx/dpu/dpu-disengcfg.c
+++ b/drivers/gpu/imx/dpu/dpu-disengcfg.c
@@ -1,6 +1,6 @@
 /*
  * Copyright (C) 2016 Freescale Semiconductor, Inc.
- * Copyright 2017 NXP
+ * Copyright 2017-2018 NXP
  *
  * This program is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License as published by the
@@ -93,12 +93,12 @@ struct dpu_disengcfg *dpu_dec_get(struct dpu_soc *dpu, int id)
 	mutex_lock(&dec->mutex);
 
 	if (dec->inuse) {
-		dec = ERR_PTR(-EBUSY);
-		goto out;
+		mutex_unlock(&dec->mutex);
+		return ERR_PTR(-EBUSY);
 	}
 
 	dec->inuse = true;
-out:
+
 	mutex_unlock(&dec->mutex);
 
 	return dec;
diff --git a/drivers/gpu/imx/dpu/dpu-extdst.c b/drivers/gpu/imx/dpu/dpu-extdst.c
index ab79fe6..09a36a9 100644
--- a/drivers/gpu/imx/dpu/dpu-extdst.c
+++ b/drivers/gpu/imx/dpu/dpu-extdst.c
@@ -1,6 +1,6 @@
 /*
  * Copyright (C) 2016 Freescale Semiconductor, Inc.
- * Copyright 2017 NXP
+ * Copyright 2017-2018 NXP
  *
  * This program is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License as published by the
@@ -418,12 +418,12 @@ struct dpu_extdst *dpu_ed_get(struct dpu_soc *dpu, int id)
 	mutex_lock(&ed->mutex);
 
 	if (ed->inuse) {
-		ed = ERR_PTR(-EBUSY);
-		goto out;
+		mutex_unlock(&ed->mutex);
+		return ERR_PTR(-EBUSY);
 	}
 
 	ed->inuse = true;
-out:
+
 	mutex_unlock(&ed->mutex);
 
 	return ed;
diff --git a/drivers/gpu/imx/dpu/dpu-fetchdecode.c b/drivers/gpu/imx/dpu/dpu-fetchdecode.c
index cbf4ca9..b329ab6 100644
--- a/drivers/gpu/imx/dpu/dpu-fetchdecode.c
+++ b/drivers/gpu/imx/dpu/dpu-fetchdecode.c
@@ -664,12 +664,12 @@ struct dpu_fetchunit *dpu_fd_get(struct dpu_soc *dpu, int id)
 	mutex_lock(&fu->mutex);
 
 	if (fu->inuse) {
-		fu = ERR_PTR(-EBUSY);
-		goto out;
+		mutex_unlock(&fu->mutex);
+		return ERR_PTR(-EBUSY);
 	}
 
 	fu->inuse = true;
-out:
+
 	mutex_unlock(&fu->mutex);
 
 	return fu;
diff --git a/drivers/gpu/imx/dpu/dpu-fetcheco.c b/drivers/gpu/imx/dpu/dpu-fetcheco.c
index 903fddc..2b851f1 100644
--- a/drivers/gpu/imx/dpu/dpu-fetcheco.c
+++ b/drivers/gpu/imx/dpu/dpu-fetcheco.c
@@ -328,12 +328,12 @@ struct dpu_fetchunit *dpu_fe_get(struct dpu_soc *dpu, int id)
 	mutex_lock(&fu->mutex);
 
 	if (fu->inuse) {
-		fu = ERR_PTR(-EBUSY);
-		goto out;
+		mutex_unlock(&fu->mutex);
+		return ERR_PTR(-EBUSY);
 	}
 
 	fu->inuse = true;
-out:
+
 	mutex_unlock(&fu->mutex);
 
 	return fu;
diff --git a/drivers/gpu/imx/dpu/dpu-fetchlayer.c b/drivers/gpu/imx/dpu/dpu-fetchlayer.c
index 15738a0..7462413 100644
--- a/drivers/gpu/imx/dpu/dpu-fetchlayer.c
+++ b/drivers/gpu/imx/dpu/dpu-fetchlayer.c
@@ -218,12 +218,12 @@ struct dpu_fetchunit *dpu_fl_get(struct dpu_soc *dpu, int id)
 	mutex_lock(&fu->mutex);
 
 	if (fu->inuse) {
-		fu = ERR_PTR(-EBUSY);
-		goto out;
+		mutex_unlock(&fu->mutex);
+		return ERR_PTR(-EBUSY);
 	}
 
 	fu->inuse = true;
-out:
+
 	mutex_unlock(&fu->mutex);
 
 	return fu;
diff --git a/drivers/gpu/imx/dpu/dpu-fetchwarp.c b/drivers/gpu/imx/dpu/dpu-fetchwarp.c
index 673c738..2e0a90a 100644
--- a/drivers/gpu/imx/dpu/dpu-fetchwarp.c
+++ b/drivers/gpu/imx/dpu/dpu-fetchwarp.c
@@ -218,12 +218,12 @@ struct dpu_fetchunit *dpu_fw_get(struct dpu_soc *dpu, int id)
 	mutex_lock(&fu->mutex);
 
 	if (fu->inuse) {
-		fu = ERR_PTR(-EBUSY);
-		goto out;
+		mutex_unlock(&fu->mutex);
+		return ERR_PTR(-EBUSY);
 	}
 
 	fu->inuse = true;
-out:
+
 	mutex_unlock(&fu->mutex);
 
 	return fu;
diff --git a/drivers/gpu/imx/dpu/dpu-framegen.c b/drivers/gpu/imx/dpu/dpu-framegen.c
index 981aefe..9f28c13 100644
--- a/drivers/gpu/imx/dpu/dpu-framegen.c
+++ b/drivers/gpu/imx/dpu/dpu-framegen.c
@@ -529,12 +529,12 @@ struct dpu_framegen *dpu_fg_get(struct dpu_soc *dpu, int id)
 	mutex_lock(&fg->mutex);
 
 	if (fg->inuse) {
-		fg = ERR_PTR(-EBUSY);
-		goto out;
+		mutex_unlock(&fg->mutex);
+		return ERR_PTR(-EBUSY);
 	}
 
 	fg->inuse = true;
-out:
+
 	mutex_unlock(&fg->mutex);
 
 	return fg;
diff --git a/drivers/gpu/imx/dpu/dpu-hscaler.c b/drivers/gpu/imx/dpu/dpu-hscaler.c
index d9e34bc..4edd9cd 100644
--- a/drivers/gpu/imx/dpu/dpu-hscaler.c
+++ b/drivers/gpu/imx/dpu/dpu-hscaler.c
@@ -1,5 +1,5 @@
 /*
- * Copyright 2017 NXP
+ * Copyright 2017-2018 NXP
  *
  * This program is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License as published by the
@@ -316,12 +316,12 @@ struct dpu_hscaler *dpu_hs_get(struct dpu_soc *dpu, int id)
 	mutex_lock(&hs->mutex);
 
 	if (hs->inuse) {
-		hs = ERR_PTR(-EBUSY);
-		goto out;
+		mutex_unlock(&hs->mutex);
+		return ERR_PTR(-EBUSY);
 	}
 
 	hs->inuse = true;
-out:
+
 	mutex_unlock(&hs->mutex);
 
 	return hs;
diff --git a/drivers/gpu/imx/dpu/dpu-layerblend.c b/drivers/gpu/imx/dpu/dpu-layerblend.c
index c184209..718c8db 100644
--- a/drivers/gpu/imx/dpu/dpu-layerblend.c
+++ b/drivers/gpu/imx/dpu/dpu-layerblend.c
@@ -329,12 +329,12 @@ struct dpu_layerblend *dpu_lb_get(struct dpu_soc *dpu, int id)
 	mutex_lock(&lb->mutex);
 
 	if (lb->inuse) {
-		lb = ERR_PTR(-EBUSY);
-		goto out;
+		mutex_unlock(&lb->mutex);
+		return ERR_PTR(-EBUSY);
 	}
 
 	lb->inuse = true;
-out:
+
 	mutex_unlock(&lb->mutex);
 
 	return lb;
diff --git a/drivers/gpu/imx/dpu/dpu-tcon.c b/drivers/gpu/imx/dpu/dpu-tcon.c
index a0a94ce..0e66a94 100644
--- a/drivers/gpu/imx/dpu/dpu-tcon.c
+++ b/drivers/gpu/imx/dpu/dpu-tcon.c
@@ -1,6 +1,6 @@
 /*
  * Copyright (C) 2016 Freescale Semiconductor, Inc.
- * Copyright 2017 NXP
+ * Copyright 2017-2018 NXP
  *
  * This program is free software; you can redistribute it and/or modify it
  * under the terms of the GNU General Public License as published by the
@@ -207,12 +207,12 @@ struct dpu_tcon *dpu_tcon_get(struct dpu_soc *dpu, int id)
 	mutex_lock(&tcon->mutex);
 
 	if (tcon->inuse) {
-		tcon = ERR_PTR(-EBUSY);
-		goto out;
+		mutex_unlock(&tcon->mutex);
+		return ERR_PTR(-EBUSY);
 	}
 
 	tcon->inuse = true;
-out:
+
 	mutex_unlock(&tcon->mutex);
 
 	return tcon;
diff --git a/drivers/gpu/imx/dpu/dpu-vscaler.c b/drivers/gpu/imx/dpu/dpu-vscaler.c
index d58042b..2f851f0 100644
--- a/drivers/gpu/imx/dpu/dpu-vscaler.c
+++ b/drivers/gpu/imx/dpu/dpu-vscaler.c
@@ -365,12 +365,12 @@ struct dpu_vscaler *dpu_vs_get(struct dpu_soc *dpu, int id)
 	mutex_lock(&vs->mutex);
 
 	if (vs->inuse) {
-		vs = ERR_PTR(-EBUSY);
-		goto out;
+		mutex_unlock(&vs->mutex);
+		return ERR_PTR(-EBUSY);
 	}
 
 	vs->inuse = true;
-out:
+
 	mutex_unlock(&vs->mutex);
 
 	return vs;
-- 
1.7.9.5

